tools/jvm/src/main/scala/org/scalajs/core/tools/linker/backend/closure/ClosureLinkerBackend.scala

```diff
@@ -75,21 +75,51 @@ final class ClosureLinkerBackend(
     module.add(new CompilerInput(ast, ast.getInputId(), false))
 
     // Compile the module
-    val closureExterns =
-      toClosureSource(ClosureLinkerBackend.ScalaJSExternsFile)
+    val closureExterns = List(
+        toClosureSource(ClosureLinkerBackend.ScalaJSExternsFile),
+        toClosureSource(makeExternsForExports(unit)))
     val options = closureOptions(output.name)
     val compiler = closureCompiler(logger)
 
     val result = logger.time("Closure: Compiler pass") {
       compiler.compileModules(
-          List(closureExterns).asJava, List(module).asJava, options)
+          closureExterns.asJava, List(module).asJava, options)
     }
 
     logger.time("Closure: Write result") {
       writeResult(result, compiler, output)
     }
   }
 
+  /** Constructs an externs file listing all exported properties in a linking
+   *  unit.
+   *
+   *  This is necessary to avoid #2491.
+   */
+  private def makeExternsForExports(linkingUnit: LinkingUnit): VirtualJSFile = {
+    import org.scalajs.core.ir.Trees._
+
+    def collectAllExportedPropertyNames(): List[String] = {
+      val allNames = for {
+        classDef <- linkingUnit.classDefs
+        exportedM <- classDef.exportedMembers
+      } yield {
+        (exportedM.tree: @unchecked) match {
+          case MethodDef(_, StringLiteral(name), _, _, _) => name
+          case PropertyDef(StringLiteral(name), _, _, _)  => name
+        }
+      }
+      allNames.filter(isValidIdentifier).distinct
+    }
+
+    val content = new java.lang.StringBuilder
+    for (exportedPropertyName <- collectAllExportedPropertyNames())
+      content.append(s"Object.prototype.$exportedPropertyName = 0;\n")
```

Comment:
Just to test my understanding:

We can't set this on the real prototype of the class, because:

- It is not visible in this scope.
- It is a total abuse of the externs feature.